### PR TITLE
feat: add `last_vote` to leaderboard

### PR DESCRIPTION
**Pull request by wa0x6e**
Add `last_vote` column to leaderboard

## Changes 

Run the following queries to add the new column and index

```sql
ALTER TABLE leaderboard ADD COLUMN last_vote BIGINT;
ALTER TABLE leaderboard ADD INDEX last_vote (`last_vote`);
```

## Test 

- Run the `refresh_leaderboard_counters` script
- It should update the `last_vote` field with the correct timestamp 
- Boot a local instance of the UI, connected to this branch
- Vote on a proposal
- The leaderboard should be updated with the correct timestamp (confirm in the DB directly)

### CODE REVIEW

File: src/helpers/actions.ts [hunk 1]
```diff
@@ -76,13 +76,13 @@ export function refreshVotesCount(spaces: string[]) {
     `
       INSERT INTO leaderboard (vote_count, user, space)
         (SELECT * FROM (
-          SELECT COUNT(votes.id) AS vote_count, voter, space
+          SELECT COUNT(votes.id) AS vote_count, MAX(created) as last_vote, voter, space
```
Comment (ChaituVR) [inline src/helpers/actions.ts:79]: while running the script i get this error 

```bash
OperationalError: ER_NON_UNIQ_ERROR: Column 'created' in field list is ambiguous
    at Query.Sequence._packetToError (/snapshot-labs/snapshot-sequencer/
```

**Review by ChaituVR — CHANGES_REQUESTED** [review]

File: src/helpers/actions.ts (hunk 1 above)
Comment (ChaituVR) [reply src/helpers/actions.ts:79]: If i change it to `votes.created` then i get `OperationalError: ER_WRONG_VALUE_COUNT_ON_ROW: Column count doesn't match value count at row 1`

Comment (ChaituVR) [reply src/helpers/actions.ts:79]: what will be the value of last_vote if there are no votes for a address

Comment (wa0x6e) [reply src/helpers/actions.ts:79]: > what will be the value of last_vote if there are no votes for a address

NULL

Comment (ChaituVR) [reply src/helpers/actions.ts:79]: Any idea how to solve the above? :(

Comment (wa0x6e) [reply src/helpers/actions.ts]: SQL error fixed

File: src/writer/vote.ts [hunk 2]
```diff
@@ -178,11 +178,11 @@ export async function action(body, ipfs, receipt, id, context): Promise<void> {
     await db.queryAsync(
       `
         INSERT INTO votes SET ?;
-        INSERT INTO leaderboard (space, user, vote_count)
-          VALUES(?, ?, 1)
-          ON DUPLICATE KEY UPDATE vote_count = vote_count + 1
+        INSERT INTO leaderboard (space, user, vote_count, last_vote)
+          VALUES(?, ?, 1, ?)
+          ON DUPLICATE KEY UPDATE vote_count = vote_count + 1, last_vote = ?
```
Comment (ChaituVR) [inline src/writer/vote.ts:183]: Maybe we should also update leaderboard while updating existing vote? if I update choice it still use timestamp of first vote

Comment (wa0x6e) [reply src/writer/vote.ts:183]: Yes, I forgot about the update vote scenario

Comment (wa0x6e) [reply src/writer/vote.ts:187]: Added

**Review by ChaituVR — APPROVED** [review]
tAck

**Pull request merged**